[ETHEREUM-CONTRACTS] ERC20x/FlowNFTs

0xdavinchee opened the pull request:
## Context

This is the first iteration of ERC20x, it includes only the CFAv1 NFT contracts and transferability is currently turned off for **both** inflow AND outflow NFTs.

Many peripheral files are impacted and therefore the changeset is far greater than just the new NFT contracts added.

## Summary of Changes 

### Protocol Contracts
#### New Contracts
##### NFTs
- `CFAv1NFTBase.sol` 
  - is an abstract base contract for the CFA NFTs and holds shared state and NFT functions (IERC721Metadata) for outflow/inflow NFT contracts
  - Is `UUPSProxiable` - see `SuperToken.sol` for a refresher if needed as it follows the same pattern
- `ConstantInflowNFT.sol`
  - Inflow NFT holds no additional storage on top of base, mainly used for emitting event and representing ownership (in this PR)
- `ConstantOutflowNFT.sol`
  - Outflow NFT holds the storage of ownership which is referenced in the inflow NFT, emits events
  - contains hook functions called in `ConstantFlowAgreementV1.sol`

#### Existing Contracts
- `ConstantFlowAgreementV1.sol`
  - No more try/catch and using the `ConstantOutflowNFT` onCreate/Update/Delete hooks
  - We also ensure these functions don't fail by checking no existence (onCreate) and existence (onUpdate/onDelete)
- `SuperToken.sol`
  - SuperToken is the parent of the NFTs: immutable logic contract addresses are set in constructor, proxies are deployed in initialize (for new SuperToken's) for new super token creation or `updateCode` for existing super tokens.
  - New storage variables for outflow nft, inflow nft, pool admin nft and pool member nft proxy contract addresses: 4 reserve slots used
  - `SuperTokenMock` contract modified (validateStorageLayout) to reflect this
- `SuperTokenFactory.sol`
  - Uncommented the `castrate` function in the `constructor`, see [here](https://github.com/superfluid-finance/protocol-monorepo/pull/1274/files#diff-8d32b78f9249c83c353fe27aa7e0ec26a75a097cbf530d1c7665e7a08cd6158cL72-L78) for the explanation why it was commented out in the first place
- Delete `CFAHookMocks.sol` because we no longer are using the cfa hook nft's

#### New Contracts (cont.)

##### Production Libs
- `SuperfluidNFTDeployerLibrary` created for deploying NFT contracts in `SuperToken` contract

##### Solidity Testing File Helpers
- `SuperfluidFrameworkDeployer.sol`
  - Extracted SuperToken deployment logic to `SuperTokenDeployer.sol`
- `SuperTokenDeployer.sol`
  - Handles SuperToken deployments for testing
> Note: this broke existing tests wherever `SuperfluidFrameworkDeployer.sol` was used, fixes made. Also, these changes were made because of the code size too large error.

### Scripts
- `deploy-test-framework.js`/`deployContractsAndToken.js`
  - script modified to work given new `SuperTokenDeployer`
> Note: this does **NOT** break existing workflows that only rely on the JS dev-scripts
- `deploy-framework.js`
  - script modified to deploy NFT logic contracts when running tests (new leaf pattern), a refresher if needed of the new leaf pattern:
    - deploy the logic contracts of NFT contracts, then pass it into SuperToken constructor to set as immutable variable

### JS Tests
- Had to change up a few things to get the tests able to run without breaking
  - Linking of NFTDeployerLibrary in the relevant files
  - Removing of expecting transfer events in cfa.behavior.ts for liquidations, this broke the tooling
  - fixing up cfav1forwarder, supertokenv1library tests slightly as well

### Solidity Tests
- Added a bunch of new tests for the new NFT contracts
- Upgradability tests written w/ Foundry
- Forked upgrade test added as well
- Refactored some of the existing test harnesses slightly (`FoundrySuperfluidTester`)

### Automation Contracts
The automation contracts had to be modified because we split the `SuperfluidFrameworkDeployer` and `SuperTokenDeployer` into two separate contracts.
However, I noticed that the test files in `autowrap` and `scheduler` both had contracts inheriting from their respective `SuperfluidTester` contracts and abstracted state and setup shared between the different tests to the `SuperfluidTester` file.
> Note: Code addition occurs in `SuperfluidTester` files and code deletion occurs in the test files that inherit from it.

### Misc
- `ethereum-contracts/package.json`: added foundry coverage and oz/contracts-upgradeable
- SDK-Core operations test: the test was not doing what I intended and it was flaky
- `deployNFTContracts` added to `ethereum-contracts/TestEnvironment.ts`
- `tsconfig.test.json` added `hardhat.config.ts` to `"include"` for vs code syntax highlighting
- `yarn.lock` updated yarn lock because of new packages

0xdavinchee committed:
CFANFTBase COFNFT WIP

- CFANFTBase base contract with shared NFT functions implemented w/ reference from OZ ERC721
- COFNFT basic NFT functionality implemented
- Installed contracts-upgradeable and hardhat-upgrades

0xdavinchee committed:
NFT implementation WIP

- Add interface files for NFT contracts
- Rename CFANFTBase => CFAv1NFTBase
- Move shared functions from NFT contracts to CFAv1NFTBase
- Rename NFT contracts
- WIP ConstantInflowNFT
- Add NFT contracts to SuperToken storage
- Add tests for new storage
- Add new storage variable external view functions in ISuperToken.sol

0xdavinchee committed:
NFT  contracts upgradability

- use OZ Upgradeable interface mainly for explicit semantics for developers
- add upgradability to CFAv1NFTBase with UUPSProxiable contract
- add proxiableUUID function in CFAOutflow/CFAInflowNFT's
- add updateCode function into CFAv1NFTBase

0xdavinchee committed:
minor cleanups

0xdavinchee committed:
constant outflow nft tests

- added onlyConstantInflowNFT modifier
- added dev-forge in ethereum-contracts
- cfav1nft base abstract contract inherited by nft tests
  - does initialization
  - has some tests
  - holds logic which will be used by both tests
- constant outflow nft basic tests (no transfer)

0xdavinchee committed:
CIF NFT tests WIP

- Added symmetrical (to COF NFT) tests to CIF NFT where relevant
- Add assume helper function
- Add event assertion helpers
- New syntax for passing tests: test_Passing_Description

0xdavinchee committed:
upgradability test wip

* upgradability tests WIP
* naming of some test functions to indicate fuzzing
* add metadata trigger function

0xdavinchee committed:
Upgradability tests

- Add some more documentation re: upgradability
- Rename test names to be more readable
- Reorder mock contracts for easier readability flow for reviewers
- Remaining upgradability tests

0xdavinchee committed:
cfa nft integration

- outflow transfer reverts
- cfa integration (no tests yet)

0xdavinchee committed:
fix tests w/ cfa integration

- remove cfa hook mock contracts
- remove cfa hook code from cfa (except for storage)
- fix up SuperfluidFrameworkDeployer (add cfa nft contract deploy + initialization)
- add deployment of cfa nft logic contracts in deploy-framework only in protocolReleaseVersion="test"
- add deployment of cfa nft proxy, initialization and attach to supertoken in deploy-super-token only in protocolReleaseVersion="test"
- try/catch expect event in ConstantFlowAgreementV1.behavior.ts due to test tooling issues...
- fix up cfav1 forwarder and supertokenv1library cfa tests

0xdavinchee committed:
rename flowDataBySenderReceiver

- rename flowDataBySenderReceiver to flowDataByTokenId
- add tokenURI to base NFT contract

0xdavinchee committed:
cfa integration tests super happy path

- test nft minted on flow create
- test nft emits metadataupdate on flow update
- test nft deleted on flow delete
- function renamed to follow consistent_Naming_Pattern
- add emitting address parameter to event asserts

0xdavinchee committed:
cleanup

0xdavinchee committed:
comment review WIP

- renaming of IDA stuff
- change from _gap to _reserveN
- delete expect Transfer in behavior
- move integration tests to outflow NFT test file (removes duplication)
- add test_Revert_Create_Flow_Overflows_When_Timestamp_Greater_Than_Uint32_Max test
- fix up mocks

0xdavinchee committed:
fix build and cleanup

- delete transfer code from constant inflow nft
- delete mint/burn code from constant outflow nft
- link nft deployer library in deploy-test-framework.js
- add foundry coverage to test-coverage script
- refactor FoundrySuperfluidTester slightly
- add missing test cases for 100% coverage

0xdavinchee committed:
add getFlow to SuperToken

0xdavinchee committed:
split deployers

- extract super token deployment from SuperfluidFrameworkDeployer to SuperTokenDeployer
- extract super token deployment tests to new test file
- create deployer base test contract for shared foundry test initialization logic
- fix deploy-test-framework and deployContractsAndToken to work with

0xdavinchee committed:
Merge branch 'dev' into erc20x

0xdavinchee committed:
NFT proxy deployment

- add in the NFT proxy deployment in SuperTokenFactory
- add in missing events for NFT logic creation and nft proxy contract creation
- add a deployer superfluid NFT deployer library
- modify validateStorageLayout for SuperTokenFactory because of new slots for logic contracts
- add in missing assume in tests
- modify SuperTokenDeployer to leverage new SuperTokenFactory.deployNFTProxyContractsAndInitialize

0xdavinchee committed:
SuperfluidNFTDeployer added

- fix tests to handle SuperfluidNFTDeployer

0xdavinchee committed:
fix up automations tests

- move shared logic and testing state to SuperfluidTester in both autowrap and scheduler tests
- fix tests because super tokens are deployed via SuperTokenDeployer now

0xdavinchee committed:
clean up ci.canary comment

0xdavinchee committed:
fix breaking test

0xdavinchee committed:
fix up tsconfigs

0xdavinchee committed:
fixes + cleanup

- install foundry where we are doing coverage tests
- add initialize to IConstantOutflow, IConstantInflow and remove inheritance from ICFAv1NFTBase
- change up number in operation.test.ts for gas estimation

0xdavinchee committed:
fix the sdk-core build

0xdavinchee committed:
should use same value

- we are deploying a new contract each time

0xdavinchee committed:
liquidations test WIP

- reward address added to SuperfluidFrameworkDeployer
- liquidation tests added (pic, pleb, pirate periods)

0xdavinchee commented on packages/ethereum-contracts/contracts/superfluid/SuperToken.sol:
```diff
@@ -714,6 +721,52 @@ contract SuperToken is
+            uint256 owedDeposit
+        )
+    {
+        IConstantFlowAgreementV1 cfaV1 = IConstantFlowAgreementV1(
```
**Medium**: Just want to double check if this is future proof enough once we add the GDA (the params taken at least).

hellwolf replied:
`+ GDA.getFlow()`; ? Also depending on if we make receiver either a pool or non-pool (not both at the same time).

0xdavinchee committed:
cleanup

0xdavinchee committed:
cleanup continued

- FlowData => CFAv1NFTFlowData
- remove `_` prefix from params
- clean up interface stuff

0xdavinchee committed:
cleanup continued

- onCreate/Update/Delete all take flowSender, flowReceiver for consistency
- getTokenId helper function to compute NFT id's with flowSender/flowReceiver

0xdavinchee committed:
cleanup continued

- interface cleanups
- cache cfaV1 in storage during initialization for CFAv1NFT contracts
- access control onCreate/Update/Delete hooks to only be called by CFAv1
- add tests to ensure cfaV1 properly initialized
- add tests to ensure revert occurs when attempting to call hook functions from outside of CFAv1
- tidy up upgradability mock contracts (given new storage variable)
- call validateStorageLayout for one of the upgraded contract tests

0xdavinchee committed:
remove SuperfluidDevNFTDeployer

0xdavinchee committed:
permission control mint/burn

- only allow COF NFT to call mint/burn on CIF NFT
- fix up bad assume logic in test

0xdavinchee committed:
add new updateLogicContracts

0xdavinchee committed:
add sanity check test

- add mock contract
- add test to ensure new code is being run

0xdavinchee committed:
Merge branch 'supertokenfactory-new-update' into erc20x

0xdavinchee committed:
move creation to helpers

0xdavinchee committed:
migration test and gas test

- full erc20x migration test added
- gas tests added
- getConstantOutflowNFTLogic() and getConstantInflowNFTLogic() added to SuperTokenFactory
- lib/forge-std updated (forge update)
- remove onlySelf modifier

0xdavinchee committed:
forge install: forge-std

v1.3.0

0xdavinchee committed:
forge install: forge-std

v1.3.0

0xdavinchee committed:
SuperTokenFactory deploy test

0xdavinchee committed:
fix framework test

0xdavinchee committed:
forge install: forge-std

v1.3.0

0xdavinchee commented on packages/ethereum-contracts/dev-scripts/deployContractsAndToken.js:
```diff
@@ -7,10 +7,10 @@ const {
     const [Deployer] = await ethers.getSigners();
 
-    const deployer = await deployTestFramework();
+    const {frameworkDeployer: deployer, superTokenDeployer} = await deployTestFramework();
```
kebab-case

hellwolf replied:
call it signer?

0xdavinchee committed:
fix redundant naming

0xdavinchee committed:
Ops flow cleanup

- delete updateLogicContracts() function
- delete SuperTokenFactoryHelper
- delete _updateSuperTokenLogic function (future updates will no longer follow this path)
- use SuperTokenDeployerLibrary.deploySuperTokenLogic to deploy non upgradeable super token
- fix breaking tests
- create some abstracted functions for deploy external library and linking

0xdavinchee committed:
cleanup

0xdavinchee committed:
Merge branch 'dev' into supertokenfactory-new-update

0xdavinchee committed:
Merge branch 'supertokenfactory-new-update' of github.com:superfluid-finance/protocol-monorepo into supertokenfactory-new-update

0xdavinchee committed:
fork baseline

- add fork baseline tests file
- rename fork super token factory upgrade test file

0xdavinchee committed:
fork baseline wip

- refactor ForkBaseline: it is responsible for loading the forked framework
- add more tests in ForkBaseline
- do not run fork tests in the CI

0xdavinchee committed:
Update CHANGELOG.md

0xdavinchee committed:
Initial cleanup after discussion

- remove try/catch in ConstantFlowAgreementV1
- rename CFAv1NFTMocks to CFAv1NFTUpgradabilityMocks
- create seperate CFAv1NFTMock.t.sol file
- Move NFT deployment stuff to CFAv1NFTBase (this will change again)

0xdavinchee committed:
Merge branch 'supertokenfactory-new-update' into erc20x

0xdavinchee committed:
fix build

0xdavinchee committed:
further cleanup

- rename initializeNFTContracts to setNFTProxyContracts
- add immutable _constantOutflowNFTLogic and _constantInflowNFTLogic
- deprecate non-upgradeable path in SuperTokenFactory
- fix up build

0xdavinchee committed:
Merge branch 'dev' into supertokenfactory-new-update

0xdavinchee committed:
deprecate non upgradeable deploy

0xdavinchee committed:
merge dev

0xdavinchee committed:
Merge branch 'supertokenfactory-new-update' into erc20x

0xdavinchee committed:
fix broken build

0xdavinchee committed:
bump solidity version

0xdavinchee committed:
deploy nft in SuperToken

- deploy, initialize and set NFT in supertoken.initialize for new super token path
- vm.assume(caller != address(sf.cfa)) added in test_Fuzz_Revert_If_On_Update_Is_Not_Called_By_CFAv1

0xdavinchee committed:
refactor and skip tests in CI

- refactor fork test contracts
- skip fork/gas tests

0xdavinchee committed:
fix build and cleanup

- fixing the js tests given new flow

0xdavinchee committed:
do existence check

- existence check required

0xdavinchee committed:
Merge branch 'dev' into erc20x

0xdavinchee committed:
additional fixes for merge

0xdavinchee committed:
SuperToken NFT deployAndSet

- setNFTProxyContracts removed on SuperToken
- setNFTProxyContracts added to SuperTokenMock
- deployAndSetNFTProxyContracts added to SuperToken
- rename immutable variables on SuperToken
- fix deploy-framework.js given new SuperToken
- remove code from deploy-super-token.js
- fix SuperTokenV1Library.CFA.test.ts
- rename ForkTester
- move stuff from FoundrySuperfluidTester back to CFAv1NFTBase
- fix up NFT tests to use superTokenMock

0xdavinchee committed:
Merge branch 'dev' into erc20x

0xdavinchee committed:
remove only test in truffle param

add in proper replacements for super token factory logic

0xdavinchee commented on packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol:
```diff
@@ -459,26 +461,14 @@ contract ConstantFlowAgreementV1 is
+            address(flowVars.token)
+        ).constantOutflowNFT();
+
+        // check for existence if NFT exists, if it does, we don't execute hook
```
**High**:  New `onCreate` hook via `ConstantOutflowNFT` - attempts to mint. Must NOT revert.

0xdavinchee commented on packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol:
```diff
@@ -509,27 +499,9 @@ contract ConstantFlowAgreementV1 is
-                }
-            }
-        }
+        IConstantOutflowNFT(
```
**High**:  New `onUpdate` hook via `ConstantOutflowNFT` - attempts to emit event. Must NOT revert.

0xdavinchee commented on packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol:
```diff
@@ -641,25 +613,15 @@ contract ConstantFlowAgreementV1 is
-                    revert CFA_HOOK_OUT_OF_GAS();
-                }
-            }
+        IConstantOutflowNFT constantOutflowNFT = IConstantOutflowNFT(
```
**High**:  New `onDelete` hook via `ConstantOutflowNFT` - attempts to burn. Must NOT revert.

0xdavinchee commented on packages/ethereum-contracts/contracts/libs/SuperfluidNFTDeployerLibrary.sol:
```diff
@@ -0,0 +1,51 @@
+import { IConstantOutflowNFT } from "../superfluid/ConstantOutflowNFT.sol";
+import { IConstantInflowNFT } from "../superfluid/ConstantInflowNFT.sol";
+
+library SuperfluidNFTDeployerLibrary {
```
**Medium**: External library used for deploying NFT proxies in `SuperToken.sol`

hellwolf replied:
okay

0xdavinchee commented on packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperToken.sol:
```diff
@@ -503,6 +514,107 @@ interface ISuperToken is ISuperfluidToken, TokenInfo, IERC20, IERC777 {
+
+    function constantOutflowNFT() external view returns (IConstantOutflowNFT);
+    function constantInflowNFT() external view returns (IConstantInflowNFT);
+    function poolAdminNFT() external view returns (IPoolAdminNFT);
```
**High**: I am making an assumption here that the Pool NFTs will be linked on the SuperToken.

0xdavinchee commented on packages/ethereum-contracts/contracts/mocks/SuperTokenFactoryMock.sol:
```diff
@@ -33,12 +33,6 @@ contract SuperTokenFactoryStorageLayoutTester is SuperTokenFactoryBase {
         assembly { slot := _canonicalWrapperSuperTokens.slot offset := _canonicalWrapperSuperTokens.offset }
         require(slot == 1 && offset == 0, "_canonicalWrapperSuperTokens changed location");
     }
-
```
**Medium**: Note that I am deleting: `createSuperTokenLogic` as this pattern no longer exists after the update of the [previous PR](https://github.com/superfluid-finance/protocol-monorepo/pull/1276)

hellwolf replied:
implication: one should do a token upgrade with 1.5.1 first before doing token upgrade using this release.

0xdavinchee commented on packages/ethereum-contracts/contracts/mocks/SuperTokenMock.sol:
```diff
@@ -60,8 +67,20 @@ contract SuperTokenStorageLayoutTester is SuperToken {
+        assembly { slot:= constantInflowNFT.slot offset := constantInflowNFT.offset }
+        require (slot == 23 && offset == 0, "constantInflowNFT changed location");
+
+        assembly { slot:= poolAdminNFT.slot offset := poolAdminNFT.offset }
```
**High**: Given my assumption that the pool NFTs will be linked to a specific token I have allotted two storage slots to be added for the pool NFTs in the future.
Worst case: they remain empty if we choose not to do this and we explicitly remove them in the future.

hellwolf replied:
*four new slots

0xdavinchee commented on packages/ethereum-contracts/contracts/superfluid/CFAv1NFTBase.sol:
```diff
@@ -0,0 +1,407 @@
+    /// @notice This returns the Uniform Resource Identifier (URI), where the metadata for the NFT lives.
+    /// @dev Returns the Uniform Resource Identifier (URI) for `tokenId` token.
+    /// @return the token URI
+    function tokenURI(
```
**Medium**: This is a reuse of the previous marketing NFT url.

0xdavinchee commented on packages/ethereum-contracts/contracts/superfluid/ConstantInflowNFT.sol:
```diff
@@ -0,0 +1,103 @@
+        uint256 tokenId,
+        bytes memory // data
+    ) internal virtual override {
+        _transfer(from, to, tokenId);
```
**Low**: Mainly a note for the future, but when we add in the transfer feature, will we want: `_checkOnERC721Received`?

0xdavinchee commented on packages/ethereum-contracts/contracts/superfluid/ConstantInflowNFT.sol:
```diff
@@ -0,0 +1,103 @@
+        _burn(tokenId);
+    }
+
+    function flowDataByTokenId(
```
**High**: Note that I am referencing the storage in `ConstantOutflowNFT` from `ConstantInflowNFT`, this is safe now, but when transferability is added, we must make sure there are no cross contract view function reentrancy vulnerabilities from this function.

0xdavinchee commented on packages/ethereum-contracts/contracts/superfluid/ConstantOutflowNFT.sol:
```diff
@@ -0,0 +1,209 @@
+contract ConstantOutflowNFT is CFAv1NFTBase {
+    /// @notice A mapping from token id to CFAv1NFTFlowData = { address sender, uint32 flowStartDate, address receiver}
+    /// @dev The token id is uint256(keccak256(abi.encode(flowSender, flowReceiver)))
+    mapping(uint256 => CFAv1NFTFlowData) internal _flowDataByTokenId;
```
**High**: This is pretty much the only novel thing being done for the NFT contracts (also upgradability), but it would be good to focus extra attention on the functions surrounding ownership and access control of the NFTs.

0xdavinchee replied:
Note that this is tested pretty heavily. Retesting of basic ERC-721 functions was trivial, but is just done as good practice and for sanity checking. It also allowed further experimentation and iteration on Foundry usage.

0xdavinchee commented on packages/ethereum-contracts/contracts/superfluid/SuperToken.sol:
```diff
@@ -42,6 +46,12 @@ contract SuperToken is
     uint8 constant private _STANDARD_DECIMALS = 18;
 
+    // solhint-disable-next-line var-name-mixedcase
+    IConstantOutflowNFT immutable public CONSTANT_OUTFLOW_NFT_LOGIC;
```
**High**: New "immutable leaf" pattern applied here. See `SuperTokenFactory.sol` and how SuperToken logic is handled for a refresher. 

Reference: https://github.com/superfluid-finance/protocol-monorepo/pull/1276

hellwolf replied:
okay

0xdavinchee commented on packages/ethereum-contracts/contracts/superfluid/SuperToken.sol:
```diff
@@ -65,39 +75,66 @@ contract SuperToken is
         SuperfluidToken(host)
         // solhint-disable-next-line no-empty-blocks
     {
+        // set the immutable canonical NFT logic addresses in construction
```
**High**: On subsequent `SuperTokenLogic` updates, we must deploy NFT logic contracts first non-atomically.
The non-atomic nature is most likely okay because this will revert if the contract has already been initialized.

0xdavinchee commented on packages/ethereum-contracts/contracts/superfluid/SuperToken.sol:
```diff
@@ -109,6 +146,12 @@ contract SuperToken is
+        // initialize the proxies, pointing to the canonical NFT logic contracts
+        // set in the constructor
+        // link the deployed NFT proxies to the SuperToken
+        _deployAndSetNFTProxyContracts();
```
**High**: On creation of new `SuperToken`'s, the initialize function deploys the NFT proxies and sets them, the proxies are also initialized to the canonical NFT logic contracts of the current version of the `SuperToken` contract.

0xdavinchee commented on packages/ethereum-contracts/contracts/superfluid/SuperToken.sol:
```diff
@@ -714,6 +757,86 @@ contract SuperToken is
+     *************************************************************************/
+
+    /// @inheritdoc ISuperToken
+    function deployAndSetNFTProxyContracts()
```
**High**: This function is to be called for deploying and settings NFT proxy contracts for existing `SuperToken`'s in the wild aka the _upgrade path_.

0xdavinchee commented on packages/ethereum-contracts/contracts/superfluid/SuperToken.sol:
```diff
@@ -714,6 +757,86 @@ contract SuperToken is
+        )
+    {
+        IConstantFlowAgreementV1 cfaV1 = IConstantFlowAgreementV1(
+            address(
```
**Low**: Any optimizations we might want to do here for storing CFAv1?

0xdavinchee commented on packages/ethereum-contracts/contracts/superfluid/SuperTokenFactory.sol:
```diff
@@ -67,15 +69,13 @@ abstract contract SuperTokenFactoryBase is
-        // here 
-        // UUPSProxiable(address(_SUPER_TOKEN_LOGIC)).castrate();
+
+        UUPSProxiable(address(_SUPER_TOKEN_LOGIC)).castrate();
```
**High:** Note that this was commented out in the previous PR. 

You can look to the left as well to see the explanation or click the link below.

https://github.com/superfluid-finance/protocol-monorepo/pull/1276/files#r1105788478

0xdavinchee commented on packages/ethereum-contracts/contracts/superfluid/SuperTokenFactory.sol:
```diff
@@ -333,17 +335,4 @@ contract SuperTokenFactory is SuperTokenFactoryBase
         // solhint-disable-next-line no-empty-blocks
     {
     }
-
```
**Medium**: Note that this is deleted as the comment below indicates it should be.

0xdavinchee commented on packages/ethereum-contracts/ops-scripts/deploy-framework.js:
```diff
@@ -563,43 +575,87 @@ module.exports = eval(`(${S.toString()})({skipArgv: true})`)(async function (
+                            .padStart(64, "0"),
+                    ]
+                );
+                const superTokenLogicCodeChanged = await codeChanged(
```
**High**: Note that this is breaking the CI currently. It is always returning true.

0xdavinchee commented on packages/ethereum-contracts/ops-scripts/deploy-framework.js:
```diff
@@ -563,43 +575,87 @@ module.exports = eval(`(${S.toString()})({skipArgv: true})`)(async function (
+                    await superTokenLogic.CONSTANT_OUTFLOW_NFT_LOGIC();
+                const constantInflowNFTLogicAddress =
+                    await superTokenLogic.CONSTANT_INFLOW_NFT_LOGIC();
+                const superTokenFactoryCodeChanged = await codeChanged(
```
**High**: The `codeChanged` logic needs to be revisited, I will do some digging, but removing all the replacements works (see dev branch).

0xdavinchee commented on packages/ethereum-contracts/ops-scripts/deploy-framework.js:
```diff
@@ -563,43 +575,87 @@ module.exports = eval(`(${S.toString()})({skipArgv: true})`)(async function (
+        // or if constant outflow nft logic or constant inflow nft logic has changed
         async () => {
             let superTokenFactoryLogic;
+            // deploy constant outflow nft logic contract
```
**Medium**: Note the new deployment flow additions. This is unoptimized and can be further optimized, but optimizations should be added only after `codeChanged` is fixed.

0xdavinchee commented on packages/ethereum-contracts/test/contracts/agreements/ConstantFlowAgreementV1.behavior.ts:
```diff
@@ -376,21 +375,6 @@ export async function _shouldChangeFlow({
                         liquidationTypeData,
                     }
                 );
-
```
**Medium**: I've removed the expect emit of the transfer event here because OZ wasn't able to capture this (confused about NFT Transfer vs. ERC20 Transfer). Foundry wasn't able to capture this either, so we may want to revisit this... Although when I look at the logs emitted when running the tests, the event is actually emitted.

0xdavinchee commented on packages/ethereum-contracts/test/foundry/DeployerBase.t.sol:
```diff
@@ -0,0 +1,64 @@
+// SPDX-License-Identifier: AGPLv3
```
**Medium**: This abstraction might not be needed. It might be overkill and we could combine this with `FoundrySuperfluidTester`. This was created to keep some of the test contracts 'lighter'. Might not have been a worthy tradeoff though.

0xdavinchee commented on packages/ethereum-contracts/test/foundry/deployments/ForkPolygonERC20xCFANFTDeployment.t.sol:
```diff
@@ -0,0 +1,294 @@
+import { CFAv1Library } from "../../../contracts/apps/CFAv1Library.sol";
+import { ForkBaselineTest } from "./ForkBaseline.t.sol";
+
+/// @title ForkPolygonERC20xCFANFTDeployment
```
**Medium**: This will break until the changes from the previous PR are merged.

0xdavinchee commented on packages/ethereum-contracts/test/foundry/superfluid/nftUpgradability/CFAv1NFTUpgradability.t.sol:
```diff
@@ -0,0 +1,275 @@
+    ICFAv1NFTBaseMockErrors
+} from "./CFAv1NFTUpgradabilityMocks.sol";
+
+/// @title ConstantFAv1NFTsUpgradabilityTest
```
**High**: These upgradability tests should be reviewed a bit more.

0xdavinchee commented on packages/ethereum-contracts/test/ops-scripts/deployment.test.js:
```diff
@@ -225,143 +225,132 @@ contract("Embedded deployment scripts", (accounts) => {
             });
 
-            // TODO deployment script upgrades detection only works for truffle
-            if (process.env.IS_TRUFFLE) {
```
**Low**: Hardhat can run these tests too, I tried locally when attempting to fix the contract logic change issue.

0xdavinchee committed:
cleanup - add signed signature

0xdavinchee committed:
fix erc20x fuzz test

hellwolf commented on .github/workflows/ci.canary.yml:
```diff
@@ -137,6 +137,18 @@ jobs:
           yarn build
           yarn workspace @superfluid-finance/ethereum-contracts test-coverage
 
+      - name: Install Foundry
```
Maybe a separate pull request?

hellwolf commented on packages/ethereum-contracts/contracts/superfluid/CFAv1NFTBase.sol:
```diff
@@ -0,0 +1,407 @@
+abstract contract CFAv1NFTBase is UUPSProxiable, ICFAv1NFTBase {
+    using Strings for uint256;
+
+    string public constant BASE_URI =
```
put to the constructor please.

0xdavinchee replied:
this cannot be made immutable: `Immutable variables cannot have a non-value type.`

0xdavinchee committed:
cleanup

- undo noisy changes
- rename deployer to admin signer
- remove DeployerBase helper contract
- add code from DeployerBase to FoundrySuperfluidTester
- add check to onUpdate to only emit event if an NFT exists

0xdavinchee committed:
fix codeChanged issue

- it was due to external library address being appended in bytecode, reused same pattern used for IDA to handle this (reusing deployed library).
- we should reevaulate this, probably requires another level of checking if the library bytecode has changed

0xdavinchee committed:
Merge branch 'dev' into erc20x

0xdavinchee committed:
fix bad merge

0xdavinchee committed:
Review comments addressed

- move existence check into NFT
- rename CFAv1NFTBase => FlowNFTBase
- delete ConstantOutflowNFTLogicCreated event
- move _safeTransfer to base nft
- delete inflowTransferMint, inflowTransferBurn, onlyConstantInflowNFT
- fix upgradability tests

0xdavinchee committed:
remove mapping variable names

0xdavinchee committed:
bump solc to 0.8.19

0xdavinchee committed:
fix liquidations test

- liquidations tests fixed
- remove foundry liquidations test
- remove deployAndSetNFTProxyContracts

0xdavinchee committed:
undo bump 0.8.19

0xdavinchee committed:
don't import ethers from hardhat

0xdavinchee commented on packages/ethereum-contracts/test/foundry/FoundrySuperfluidTester.sol:
```diff
@@ -30,34 +50,60 @@ contract FoundrySuperfluidTester is Test {
+        // - Host
+        // - CFA
+        // - IDA
+        // - SuperTokenFactory
```
Add NFT logic contracts too

0xdavinchee commented on packages/ethereum-contracts/test/foundry/FoundrySuperfluidTester.sol:
```diff
@@ -30,34 +50,60 @@ contract FoundrySuperfluidTester is Test {
+        // transfer ownership of TestGovernance to superTokenDeployer
+        // governance ownership is required for initializing the NFT
+        // contracts on the SuperToken
+        sfDeployer.transferOwnership(address(superTokenDeployer));
```
this is probably unnecessary

0xdavinchee commented on packages/ethereum-contracts/test/foundry/FoundrySuperfluidTester.sol:
```diff
@@ -70,27 +116,187 @@ contract FoundrySuperfluidTester is Test {
+        int96 receiverFlowAfter = superToken.getNetFlowRate(flowReceiver);
+
+        assertEq(
+            senderFlowAfter,
```
rename to `senderNetFlowAfter`

0xdavinchee commented on packages/ethereum-contracts/test/foundry/superfluid/ConstantOutflowNFT.t.sol:
```diff
@@ -0,0 +1,572 @@
+        vm.prank(flowSender);
+        sf.cfaLib.deleteFlow(flowSender, flowReceiver, superTokenMock);
+
+        assert_Flow_Data_State_IsEmpty(nftId);
```
rename to `NFT_Flow_...`

0xdavinchee commented on packages/ethereum-contracts/test/foundry/deployments/ForkPolygonERC20xCFANFTDeployment.t.sol:
```diff
@@ -0,0 +1,294 @@
+
+        newLogic = new SuperTokenFactory(sfFramework.host, newSuperTokenLogic);
+
+        // SuperTokenFactory.updateCode
```
@note: fix this up

0xdavinchee committed:
cleanup

- remove upgradeable
- cleanup fork deployment test
- rename `senderFlowAfter` to `senderNetFlowAfter` and `receiverFlowAfter` to `receiverNetFlowAfter`
- `assert_Flow_Data` => `assert_NFT_Flow_Data`
- delete transfer ownership back to superTokenDeployer in FoundrySuperfluidTester

0xdavinchee committed:
Merge branch 'dev' into erc20x

0xdavinchee committed:
Merge branch 'dev' into erc20x

0xdavinchee committed:
Merge branch 'dev' into erc20x

0xdavinchee committed:
clean up interfaces

- clean up IConstantInflowNFT and IConstantOutflowNFT
  - inherit from IFlowNFTBase and delete functions defined in IFlowNFTBase
- add missing functions to IFlowNFTBase
- ConstantInflow and ConstantOutflowNFT inherit from respective interfaces

0xdavinchee committed:
Merge branch 'dev' into erc20x

ngmachado commented on packages/ethereum-contracts/contracts/superfluid/ConstantOutflowNFT.sol:
```diff
@@ -0,0 +1,197 @@
+        if (_exists(newTokenId)) {
+            revert COF_NFT_TOKEN_ALREADY_EXISTS();
+        }
+        if (block.timestamp != uint256(uint32(block.timestamp))) {
```
Don't know if valuable to make this check

0xdavinchee committed:
remove overflow check

overkill check!

0xdavinchee committed:
Merge branch 'dev' into erc20x

0xdavinchee committed:
comments on breakage

0xdavinchee committed:
Merge branch 'dev' into erc20x

0xdavinchee committed:
bump to 0.8.19

0xdavinchee committed:
Merge branch 'dev' into erc20x

hellwolf reviewed (approved):
LFG

kasparkallas commented on packages/ethereum-contracts/contracts/agreements/ConstantFlowAgreementV1.sol:
```diff
@@ -641,26 +614,12 @@ contract ConstantFlowAgreementV1 is
+        // for now, we are aware that this will break super tokens with custom super token logic
+        // which choose not to upgrade, in order to fix this, we must use the old try/catch logic
+        // which was previously in place for the previous marketing NFT
+        IConstantOutflowNFT(
```
Could clean this up a bit to match the syntax used elsewhere.

0xdavinchee replied:
addressed

kasparkallas commented on packages/ethereum-contracts/contracts/superfluid/ConstantInflowNFT.sol:
```diff
@@ -0,0 +1,105 @@
+    /// @param to the receiver of the inflow nft and desired flow receiver
+    /// @param newTokenId the new token id
+    function mint(
+        address to,
```
Theoretically could have done without this and just loaded ConstantOutflowNFT and taken the receiver from there.

kasparkallas commented on packages/ethereum-contracts/contracts/superfluid/FlowNFTBase.sol:
```diff
@@ -0,0 +1,389 @@
+    using Strings for uint256;
+
+    string public constant BASE_URI =
+        "https://nft.superfluid.finance/cfa/v1/getmeta";
```
Assuming this can be updated? :wink: @hellwolf  @d10r

hellwolf replied:
yes, "constant" means it's part of the logic contract code.

kasparkallas commented on packages/ethereum-contracts/contracts/superfluid/FlowNFTBase.sol:
```diff
@@ -0,0 +1,389 @@
+    ISuperToken public superToken;
+
+    string internal _name;
+    string internal _symbol;
```
Theoretically this could be calculated from the token or stored in the metadata.

hellwolf replied:
Possible. To what advantage you were thinking of?

Background info: The string lib written in solidity is not so great, but for only simple concatenation it is okay. And this piece of logic could also be updated.

What do you mean stored in the metadata?

kasparkallas replied:
Advantage? _Something-something_ gas and storage? :smile: When I wrote code then I tend to optimize towards less state and more calculated properties. Just noticed that it's possible here but I'm not sure if it's of value in this case in the Solidity-land.

Stored in the metadata? I remembered that ERC-1155 forewent storing `name` and `symbol` and only relied on the metadata: https://eips.ethereum.org/EIPS/eip-1155#metadata-choices I didn't remember that this meant breaking ERC-20 and ERC-721 standards...

kasparkallas commented on packages/ethereum-contracts/contracts/superfluid/FlowNFTBase.sol:
```diff
@@ -0,0 +1,389 @@
+    function balanceOf(
+        address // owner
+    ) external pure returns (uint256 balance) {
+        balance = 1;
```
Could theoretically use a stored counter solution here to get an actual count. But heard we didn't do it because of gas reasons.

hellwolf replied:
Also, since we can't force retro mint, the counter wouldn't reflect correct numbers for existing ones.

kasparkallas replied:
I think it would reflect the correct number for this particular NFT.

What happens to the old NFTs btw? EDIT: Got the answer that they'll keep on existing without retro burn. We'll keep the endpoint up and running. We also brought up the point of using a newer metadata URL for the new endpoints and not force it to support the old NFTs.

kasparkallas commented on packages/ethereum-contracts/contracts/superfluid/FlowNFTBase.sol:
```diff
@@ -0,0 +1,389 @@
+        string memory superTokenSymbol = superToken.symbol();
+
+        (
+            uint256 startDate,
```
This is actually `updateDate`, i.e. this is not the date flow was started at...

0xdavinchee replied:
resolved

kasparkallas commented on packages/ethereum-contracts/contracts/superfluid/FlowNFTBase.sol:
```diff
@@ -0,0 +1,389 @@
+                    "&flowRate=",
+                    uint256(uint96(flowRate)).toString(),
+                    "&start_date=",
+                    startDate.toString()
```
@d10r is it supposed to be that the date of the update goes here for the `start_date`? (The variable `startDate` is not actually the _start date_ and is _update date_ instead. Look at this comment: https://github.com/superfluid-finance/protocol-monorepo/pull/1274#discussion_r1142370439)

kasparkallas commented on packages/ethereum-contracts/contracts/superfluid/FlowNFTBase.sol:
```diff
@@ -0,0 +1,389 @@
+
+    /// @notice Mapping for token approvals
+    /// @dev tokenID => approved address mapping
+    mapping(uint256 => address) internal _tokenApprovals;
```
InflowNFT doesn't delete this but OutflowNFT does.

0xdavinchee replied:
addressed

hellwolf replied:
good catch.

0xdavinchee committed:
address review comments

- approval burn forgotten on inflow NFT added
- clean up in CFAv1

0xdavinchee committed:
Merge branch 'dev' into erc20x

0xdavinchee committed:
startDate -> lastUpdatedAt

sirpy commented on packages/ethereum-contracts/contracts/superfluid/SuperToken.sol:
```diff
@@ -2,7 +2,7 @@
 
 import { UUPSProxiable } from "../upgradability/UUPSProxiable.sol";
-
+import { IConstantFlowAgreementV1 } from "../interfaces/agreements/IConstantFlowAgreementV1.sol";
```
why was this import added?

0xdavinchee replied:
good point, these are unused, resolved

sirpy commented on packages/ethereum-contracts/contracts/superfluid/SuperToken.sol:
```diff
@@ -13,15 +13,19 @@ import {
 import { ERC777Helper } from "../libs/ERC777Helper.sol";
-
 import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import { Ownable } from "@openzeppelin/contracts/access/Ownable.sol";
```
why was this import added?

0xdavinchee replied:
good point, these are unused, resolved

kasparkallas reviewed (approved):
Approved but the TODO is to set FlowNFTBase's BASE_URI properly. It should be set up in a way where we could **detect whether the metadata request is coming for the incoming or the outgoing NFT**.

hellwolf committed:
Merge remote-tracking branch 'origin/dev' into erc20x

djmartberry66 commented:
Smartberry1122
